paimon-core/src/main/java/org/apache/paimon/mergetree/MergeTreeReaders.java

```diff
@@ -70,8 +71,14 @@ public static RecordReader<KeyValue> readerForSection(
             MergeFunctionWrapper<KeyValue> mergeFunctionWrapper)
             throws IOException {
         List<RecordReader<KeyValue>> readers = new ArrayList<>();
-        for (SortedRun run : section) {
-            readers.add(readerForRun(run, readerFactory));
+        // if one of the readers creating failed, we need to close them all.
+        try {
+            for (SortedRun run : section) {
+                readers.add(readerForRun(run, readerFactory));
```

Comment:
`readerForRun` is also used in `ChangelogMergeTreeRewriter#rewriteChangelogCompaction`. Consider creating a method `List<RecordReader<KeyValue>> readerListForSection(List<SortedRun> section, KeyValueFileReaderFactory readerFactory)`.